container/container.go

```diff
@@ -47,151 +64,409 @@ var statuses = []struct {
+		UID: os.Geteuid(),
+		GID: os.Getegid()},
+	}); err != nil {
+		return err
 	}
-	return c, nil
-}
 
-// Build builds a docker image.
-func (c *DockerContainer) Build(path string) (tag string, err error) {
 	excludeFiles, err := build.ReadDockerignore(path)
 	if err != nil {
-		return "", err
+		return err
 	}
 
 	tr, err := archive.TarWithOptions(path, &archive.TarOptions{
 		Compression:     archive.Gzip,
 		ExcludePatterns: excludeFiles,
 	})
 	if err != nil {
-		return "", err
+		return err
 	}
 	defer tr.Close()
 
-	resp, err := c.client.ImageBuild(context.Background(), tr, types.ImageBuildOptions{
+	if _, err := c.client.ImageBuild(context.Background(), tr, types.ImageBuildOptions{
 		Remove:         true,
 		ForceRemove:    true,
 		SuppressOutput: true,
-	})
+		Tags:           []string{imageTag + srv.Hash.String()},
+	}); err != nil {
+		return err
+	}
+
+	return nil
+}
+
+// Start starts the service.
+func (c *Container) Start(srv *service.Service, instanceHash, runnerHash, instanceEnvHash hash.Hash, instanceEnv []string, registerPayload []byte) (err error) {
+	runnerName := servicePrefix + runnerHash.String()
+
+	networkID, err := c.createNetwork(runnerName)
 	if err != nil {
-		return "", err
+		return err
 	}
-	defer resp.Body.Close()
-	return tagFromResponse(resp.Body)
+	enginedNetworkID, err := c.enginedNetworkID()
+	if err != nil {
+		return err
+	}
+
+	specs := make([]swarm.ServiceSpec, 0)
+
+	// Create dependency container configs
+	for _, dep := range srv.Dependencies {
+		depName := runnerName + "_" + dep.Key
+		volumes := convertVolumes(srv, dep.Volumes, dep.Key)
+		volumesFrom, err := convertVolumesFrom(srv, dep.VolumesFrom)
+		if err != nil {
+			return err
+		}
+		stopGracePeriod := defaultStopGracePeriod
+		maxAttempts := defaultMaxAttempts
+		specs = append(specs, swarm.ServiceSpec{
+			Annotations: swarm.Annotations{
+				Name: depName,
+				Labels: map[string]string{
+					"mesg.engine":                c.engineName,
+					"mesg.service":               srv.Hash.String(),
+					"mesg.instance":              instanceHash.String(),
+					"mesg.runner":                runnerHash.String(),
+					"mesg.dependency":            dep.Key,
+					"com.docker.stack.namespace": depName,
+				},
+			},
+			TaskTemplate: swarm.TaskSpec{
+				ContainerSpec: &swarm.ContainerSpec{
+					Image: dep.Image,
+					Labels: map[string]string{
+						"com.docker.stack.namespace": depName,
+					},
+					Env:             dep.Env,
+					Args:            dep.Args,
+					Command:         strings.Fields(dep.Command),
+					Mounts:          append(volumes, volumesFrom...),
+					StopGracePeriod: &stopGracePeriod,
+				},
+				Networks: []swarm.NetworkAttachmentConfig{
+					{
+						Target:  networkID,
+						Aliases: []string{dep.Key},
+					},
+				},
+				RestartPolicy: &swarm.RestartPolicy{
+					Condition:   swarm.RestartPolicyConditionOnFailure,
+					MaxAttempts: &maxAttempts,
+				},
```

Comment:
I added this policy that only restard the container 3 times. By default it's unlimited.